src/bp/ui-shared/src/AccessControl/auth.ts

```diff
@@ -0,0 +1,97 @@
+export const resourceMatches = (pattern: string, res: string) => {
+  const separator = /[\/\.]/
+  pattern = pattern || ''
+
+  if (pattern.indexOf('*') < 0) {
+    pattern = pattern + '.*'
+  }
+
+  const parts = pattern.split(separator)
+  const testParts = res.split(separator)
+
+  let matches = true
+  for (let ii = 0; matches && ii < parts.length; ii++) {
+    if (parts[ii] === '*') {
+      continue
+    } else if (ii < testParts.length) {
+      matches = parts[ii].toLowerCase() === testParts[ii].toLowerCase()
+    } else {
+      matches = false
+    }
+  }
+
+  return matches
+}
+
+const OPERATION_ALIASES = {
+  read: 'r',
+  write: 'w'
+}
+
+const KNOWN_OPERATIONS = ['r', 'w']
+
+export type Rule = {
+  op: string
+  res: string
+}
+
+export const checkRule = (rules: Rule[] | undefined | null, operation: string, resource: string) => {
+  if (!rules) {
+    return false
+  }
+
+  operation = operation.toLowerCase()
+  operation = OPERATION_ALIASES[operation] || operation
+
+  if (!KNOWN_OPERATIONS.includes(operation)) {
+    throw new Error(`Invalid rule operation: ${operation}`)
+  }
+
+  let permission = false // Everything is restricted by default
+
+  for (const rule of rules) {
```

Comment:
```suggestion
  rules.foreach(rule => {
```

Reply:
Sometimes I prefer for .. of, and since I just moved this file from common to here, i'd like to avoid having to test it extensively